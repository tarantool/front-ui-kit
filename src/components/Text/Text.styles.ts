import { css } from '@emotion/css';

import { baseFontFamily, colors, monoFontFamily } from '../../variables';

export const textStyles = {
  h1: css`
    margin: 0;
    font-family: ${baseFontFamily};
    font-size: 24px;
    line-height: 40px;
    font-weight: 600;
    color: ${colors.dark};
    -webkit-font-smoothing: antialiased;
  `,
  h2: css`
    margin: 0;
    font-family: ${baseFontFamily};
    font-size: 20px;
    line-height: 28px;
    font-weight: 600;
    color: ${colors.dark};
    -webkit-font-smoothing: antialiased;
  `,
  h3: css`
    margin: 0;
    font-family: ${baseFontFamily};
    font-size: 16px;
    line-height: 24px;
    font-weight: 600;
    color: ${colors.dark};
    -webkit-font-smoothing: antialiased;
  `,
  h4: css`
    margin: 0;
    font-family: ${baseFontFamily};
    font-size: 14px;
    line-height: 22px;
    font-weight: 600;
    color: ${colors.dark};
    -webkit-font-smoothing: antialiased;
  `,
  h5: css`
    margin: 0;
    font-family: ${baseFontFamily};
    font-size: 12px;
    line-height: 22px;
    text-transform: uppercase;
    letter-spacing: 0.01em;
    color: ${colors.dark};
    font-weight: 600;
    -webkit-font-smoothing: antialiased;
  `,
  p: css`
    margin: 0;
    font-family: ${baseFontFamily};
    font-size: 12px;
    line-height: 20px;
    color: #000000;
    -webkit-font-smoothing: antialiased;
  `,
  code: css`
    margin: 0;
    font-family: ${monoFontFamily};
    font-size: 14px;
    line-height: 20px;
    color: #000000;
    -webkit-font-smoothing: antialiased;
    b {
      font-weight: 600;
    }
  `,
  basic: css`
    margin: 0;
    font-family: ${baseFontFamily};
    font-size: 14px;
    line-height: 22px;
    color: #000000;
    -webkit-font-smoothing: antialiased;
    b {
      font-weight: 600;
    }
  `,
  upperCase: css`
    text-transform: uppercase;
  `,
  noCase: css`
    text-transform: none;
  `,
};
