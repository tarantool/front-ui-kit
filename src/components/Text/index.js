// @flow
import * as React from 'react'
import { css, cx } from 'emotion'

export const textStyles = {
  h1: css`
    margin: 0;
    font-family: 'Open Sans', sans-serif;
    font-size: 24px;
    line-height: 38px;
    font-weight: 600;
    letter-spacing: 0.48px;
    text-transform: uppercase;
    color: #000000;
  `,
  h2: css`
    margin: 0;
    font-family: 'Open Sans', sans-serif;
    font-size: 20px;
    line-height: 28px;
    font-weight: 600;
    letter-spacing: 0.4px;
    text-transform: uppercase;
    color: #000000;
  `,
  h3: css`
    margin: 0;
    font-family: 'Open Sans', sans-serif;
    font-size: 16px;
    line-height: 24px;
    font-weight: 600;
    letter-spacing: 0.32px;
    text-transform: uppercase;
    color: #cf1322;
  `,
  h4: css`
    margin: 0;
    font-family: 'Open Sans', sans-serif;
    font-size: 16px;
    line-height: 24px;
    font-weight: 600;
    letter-spacing: 0.32px;
    color: #000000;
  `,
  h5: css`
    margin: 0;
    font-family: 'Open Sans', sans-serif;
    font-size: 12px;
    line-height: 22px;
    letter-spacing: 0.24px;
    color: #000000;
  `,
  p: css`
    margin: 0;
    font-family: 'Open Sans', sans-serif;
    font-size: 12px;
    line-height: 20px;
    letter-spacing: 0.28px;
    color: #000000;
  `,
  code: css`
    margin: 0;
    font-family: 'Monaco', monospace;
    font-size: 14px;
    line-height: 22px;
    color: #000000;
    b {
      font-weight: 600;
    }
  `,
  basic: css`
    margin: 0;
    font-family: 'Open Sans', sans-serif;
    font-size: 14px;
    line-height: 22px;
    color: #000000;
    b {
      font-weight: 600;
    }
  `,
  upperCase: css`
    text-transform: uppercase;
  `,
  noCase: css`
    text-transform: none;
  `
};

type TextProps = {
  className?: string,
  children?: React.Node,
  tag?: string,
  upperCase?: boolean,
  noCase?: boolean,
  title?: string,
  variant?:
    |'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'p' | 'basic' | 'code',
  // colorVariant?: 'inherit' | 'primary' | 'secondary' | 'disabled' | 'white' | 'error',
};

export const Text = ({
  className,
  children,
  tag,
  upperCase,
  noCase,
  title,
  variant = 'basic'
}:
TextProps) => {
  const Element = tag || (variant === 'basic' ? 'span' : variant);

  return (
    <Element
      className={cx(
        textStyles[variant],
        {
          [textStyles.upperCase]: upperCase,
          [textStyles.noCase]: noCase
        },
        className
      )}
      title={title}
    >
      {children}
    </Element>
  );
};
