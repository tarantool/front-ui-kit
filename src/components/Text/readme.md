```
<Text variant='h1'>Heading 1</Text>
<Text variant='h2'>Heading 2</Text>
<Text variant='h3'>Heading 3</Text>
<Text variant='h4'>Heading 4</Text>
<Text variant='h5'>Heading 5</Text>
<Text variant='p'>paragraph</Text>
<Text>basic</Text>
<br />
<Text variant='h3' tag='span'>H3 styled span</Text>
<Text variant='h4' tag='span'>H4 styled span</Text>
```
