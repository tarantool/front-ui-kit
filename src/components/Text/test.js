import React from 'react';
import renderer from 'react-test-renderer';
import { Text } from './index';

it('Text renders correctly', () => {
  const tree = renderer.create(
    <>
      <Text variant='h1' title='Text title'>Heading 1 - 24px, 38px, 600, spacing 0.48px</Text>
      <Text variant='h2'>Heading 2 - 20px, 28px, 600, spacing 0.4px</Text>
      <Text variant='h3'>Heading 3 - 16px, 24px, 600, spacing 0.32px</Text>
      <Text variant='h4'>Heading 4 - 16px, 24px, 600, spacing 0.32px</Text>
      <Text variant='h5'>Heading 5 - 12px, 22px, 400, spacing 0.24px</Text>
      <Text variant='p'>paragraph - 12px, 20px, 400, spacing: 0.28px</Text>
      <Text>basic - 14px, 22px, 400</Text>
      <Text variant='code'>code - 14px, 22px, 400</Text>
      <Text variant='h3' tag='span'>H3 styled span</Text>
      <Text variant='h4' tag='span'>H4 styled span</Text>
    </>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
