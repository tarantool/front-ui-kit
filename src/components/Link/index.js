// @flow
import * as React from 'react';
import { css, cx } from '@emotion/css';
import { textStyles, type TextStyleVariants } from '../Text';

// TODO: Add focused state
const styles = {
  link: css`
    color: inherit;
    text-decoration: underline;

    /* &:focus {
      outline: solid 1px;
      outline-offset: 2px;
    } */

    &:hover,
    &:active {
      text-decoration: none;
    }
  `
};

type LinkProps = {
  children?: React.Node,
  className?: string,
  href: string,
  onClick?: (e: MouseEvent) => void,
  target?: string,
  title?: string,
  variant?: TextStyleVariants
};

export const Link = (
  {
    className,
    variant,
    ...props
  }: LinkProps
) => {
  const textStyle = (variant && textStyles[variant]) || null;

  /* eslint-disable jsx-a11y/anchor-has-content */
  return (
    <a
      {...props}
      className={cx(textStyle, styles.link, className)}
    />
  )
};
