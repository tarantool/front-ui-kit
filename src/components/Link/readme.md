By default link inherit text styles from parent node.
It's useful for inline links in text.

```js
import { Text } from '../Text';

<Text>Are you <Link href='/'>here</Link>?</Text>
```

But if you want to use standalone link without inherited styles,
you should pass `variant` prop.

```js
<>
  <Link href='/' variant='p'>Go there</Link>
  <br />
  <Link href='/' variant='code'>Go there</Link>
</>
```
