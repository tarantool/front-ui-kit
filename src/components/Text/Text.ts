import { MouseEvent, ReactNode, createElement, forwardRef } from 'react';
import { cx } from '@emotion/css';

import { textStyles } from './Text.styles';

export { textStyles };

export type TextStyleVariants = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'p' | 'basic' | 'code';

export interface TextProps {
  className?: string;
  htmlFor?: string;
  children?: ReactNode;
  tag?: string;
  href?: string;
  upperCase?: boolean;
  noCase?: boolean;
  title?: string;
  variant?: TextStyleVariants;
  onClick?: (event: MouseEvent<HTMLElement>) => void;
  // colorVariant?: 'inherit' | 'primary' | 'secondary' | 'disabled' | 'white' | 'error',
}

export const Text = forwardRef<HTMLElement, TextProps>(
  ({ className, children, tag, upperCase, noCase, variant = 'basic', ...props }, ref) => {
    const Element = tag || (variant === 'basic' ? 'span' : variant);

    return createElement(
      Element,
      {
        ...props,
        className: cx(
          textStyles[variant],
          {
            [textStyles.upperCase]: upperCase,
            [textStyles.noCase]: noCase,
          },
          className
        ),
        ref,
      },
      children
    );
  }
);
