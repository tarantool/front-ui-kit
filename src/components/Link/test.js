import React from 'react';
import renderer from 'react-test-renderer';

import { Link } from '../Link';
import { Text } from '../Text';

describe('Link', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <>
          <Text>
            Are you <Link href="/">here</Link>?
          </Text>
          <Link href="/" variant="p">
            Go there
          </Link>
          <Link href="/" variant="code">
            Go there
          </Link>
        </>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
