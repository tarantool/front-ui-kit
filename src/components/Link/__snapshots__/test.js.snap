// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly 1`] = `
Array [
  <span
    className="css-1ggp6x0"
  >
    Are you 
    <a
      className="css-k47rv5"
      href="/"
    >
      here
    </a>
    ?
  </span>,
  <a
    className="css-16g2h8v"
    href="/"
  >
    Go there
  </a>,
  <a
    className="css-s19i51"
    href="/"
  >
    Go there
  </a>,
]
`;
